mlir/lib/Conversion/TosaToRock/TosaToRock.cpp

```diff
@@ -300,141 +306,106 @@ class MatMulConverter final : public OpConversionPattern<tosa::MatMulOp> {
+                                ConversionPatternRewriter &b) const final {
+    Location loc = collapseOp.getLoc();
+    ArrayRef<int64_t> inpShape = collapseOp.getSrcType().getShape();
+    ArrayRef<int64_t> outShape = collapseOp.getResultType().getShape();
+    SmallVector<ReassociationIndices, 4> reassocs =
+        collapseOp.getReassociationIndices();
+
+    rock::TransformMapAttr collapseAttr =
+        rock::transformCollapseShape(b, loc, inpShape, outShape, reassocs);
+    if (!collapseAttr)
+      return b.notifyMatchFailure(
+          loc, "couldn't translate tensor collapse into rock transforms");
+    b.replaceOpWithNewOp<rock::TransformOp>(collapseOp, adaptor.getSrc(),
+                                            collapseAttr);
+    return success();
+  }
+};
+
+struct ExpandShapeRewritePattern
+    : public OpConversionPattern<tensor::ExpandShapeOp> {
+  using OpConversionPattern<tensor::ExpandShapeOp>::OpConversionPattern;
+
+  LogicalResult matchAndRewrite(tensor::ExpandShapeOp expandOp,
+                                OpAdaptor adaptor,
+                                ConversionPatternRewriter &b) const final {
+    Location loc = expandOp.getLoc();
+    ArrayRef<int64_t> inpShape = expandOp.getSrcType().getShape();
+    ArrayRef<int64_t> outShape = expandOp.getResultType().getShape();
+    SmallVector<ReassociationIndices, 4> reassocs =
+        expandOp.getReassociationIndices();
+
+    rock::TransformMapAttr expandAttr =
+        rock::transformExpandShape(b, loc, inpShape, outShape, reassocs);
+    if (!expandAttr)
+      return b.notifyMatchFailure(
+          loc, "could not translate tensor expansion into rock transform");
+    b.replaceOpWithNewOp<rock::TransformOp>(expandOp, adaptor.getSrc(),
+                                            expandAttr);
+    return success();
+  }
+};
+
 struct TransposeRewritePattern : public OpRewritePattern<tosa::TransposeOp> {
   using OpRewritePattern<tosa::TransposeOp>::OpRewritePattern;
 
   SmallVector<int32_t> getTransposeDims(Value v) const {
-    if (Operation *cval = v.getDefiningOp<arith::ConstantOp>()) {
-      auto cattr = cval->getAttr("value").cast<DenseElementsAttr>();
-      return SmallVector<int32_t>(cattr.getValues<int64_t>());
-    }
-    if (Operation *cval = v.getDefiningOp<tosa::ConstOp>()) {
+    Operation *cval = v.getDefiningOp();
+    if (isa<arith::ConstantOp>(cval) || isa<tosa::ConstOp>(cval)) {
       auto cattr = cval->getAttr("value").cast<DenseElementsAttr>();
       auto vals = cattr.tryGetValues<int32_t>();
       if (succeeded(vals))
         return SmallVector<int32_t>(*vals);
       auto vals64 = cattr.tryGetValues<int64_t>();
-      if (succeeded(vals64))
-        return SmallVector<int32_t>(*vals64);
+      assert(succeeded(vals64));
+      return SmallVector<int32_t>(*vals64);
     }
     // May be bufferization cast
     //  but this is no longer a bufferization pass, so assert
     assert(0);
```

Comment:
Can we drop this section of the function?

Also, I'd argue for this entire thing being a "fail the pass" function, not a "crash the program" one.

That is, adding non-constant transpose indices isn't a case where we've violated our own invariants, but one where the programmer passed in input this pass can't handle. So we should fail gracefully.

Reply:
yep